### PR TITLE
Cherry pick of #78495: Fix issues in kubelet for Aarch64 resulting in kubelet crashing on starup

### PR DESCRIPTION
**Pull request by odinuge**
**TL;DR**: This fixes issues in kubelet for AArch64 causing it to crash on startup. Without back porting this, the first working release for AArch64 running linux 5.0+ will land in september (v1.16.0).

When starting kubelet, it exits with 1 and prints:
```
Failed to start ContainerManager failed to initialize top level QOS containers: failed to update top level Burstable QOS cgroup : failed to set supported cgroup subsystems for cgroup [kubepods burstable]: Failed to set config for supported subsystems : failed to write 4611686018427387904 to hugetlb.64kB.limit_in_bytes: open /sys/fs/cgroup/hugetlb/kubepods/burstable/hugetlb.64kB.limit_in_bytes: permission denied
```

This is tested on `AArch64`, and works as expected: https://github.com/kubernetes/kubernetes/pull/79671#issuecomment-508585814

A bit unhappy with having to bump dependencies in a back port, but I do think it is required... This is a result of "wrong code" in k8s and runc, just something we haven't seen enabled by default before linux 5.0. :/

Just as @Pennyzct said in the original PR, #78495 (comment), the only way to run a working kubelet when running linux 5.0+ on AArch64 (The 64bits arm) is currently to compile your own version from master. No pervious releases will work (or i guess those pre hugetlb should work), and 1.16 (with the fix) will not land before mid september.

There are also people in k3s reporting this issue: rancher/k3s#474

Again, I am not a huge fan of this, and I think we at least should consider cherry picking it to 1.15, so that there is one supported version. It is also possible to back port even more, but I think doing only 1.15 is a valid choice in the middle.

For more background info, read through the (the whole discussion, not just the first post) original PR:  #78495 for all the background information.

--- 

Cherry pick of #78495 on release-1.15.

#78495: Update dependency opencontainer/runc

### CODE REVIEW

**Comment by odinuge** [comment]
/assign @derekwaynecarr 
/assign @mattjmcnaughton

Need to verify on AArch64. I don't have any hardware at hand right now.

/hold

**Comment by mattjmcnaughton** [comment]
Please ping me when you've verified on the proper hardware and I will take a look.

/ok-to-test

**Comment by odinuge** [comment]
Thanks @mattjmcnaughton!

This also depends on https://github.com/kubernetes/kubernetes/pull/78187/commits/51177c1ebbb105e354e114a56fa12cc2aee0993e. Guess i can just cherry-pick that commit (and not the whole PR) into this PR?

Since runc also depends on the newest version of `libseccomp`, imo. this PR should also be cherry-picked: https://github.com/kubernetes/kubernetes/pull/79287

**Comment by mattjmcnaughton** [comment]
Hmmm @odinuge to be honest I'm getting a little nervous about how big a cherry-pick this is (especially with all the dependency updates we need to bring in). Would you be comfortable sharing more about the impact of this bug, and how difficult you imagine it being for those impacted by this bug to upgrade to a newer version of k8s?

**Comment by odinuge** [comment]
**TL;DR**: Without back porting this, the first working release for AArch64 running linux 5.0+ will land in september (`1.16`).

Yes, ofc. @mattjmcnaughton ! 😄 

A bit unhappy with having to bump dependencies in a back port myself, but I do think it is required... This is a result of "wrong code" in k8s and runc, just something we haven't seen enabled by default before linux `5.0`.  :/

Just as @Pennyzct said in the original PR, https://github.com/kubernetes/kubernetes/pull/78495#issuecomment-507086776, the only way to run a working kubelet when running linux 5.0+ on AArch64 (The 64bits arm) is currently to compile your own version from master. No pervious releases will work (or i guess those pre hugetlb should work), and `1.16` (with the fix) will not land before mid september.

There are also people in k3s reporting this issue: https://github.com/rancher/k3s/issues/474 

Again, I am not a huge fan of this, and I think we at least should consider cherry picking it to 1.15, so that there is _one_ supported version. It is also possible to back port even more, but I think doing only 1.15 is a valid choice in the middle.

**Comment by cblecker** [comment]
Please update the title/body to reflect that this is now more than just an automated cherry pick. Exact details of what you're cherry picking should be clear.

My biggest question is when did we *announce* support for AArch64? *Have* we announced support?

**Comment by odinuge** [comment]
> Please update the title/body to reflect that this is now more than just an automated cherry pick. Exact details of what you're cherry picking should be clear.

Sorry, should have done that a long time ago.. Does the new title/desc look better?

> My biggest question is when did we announce support for AArch64? Have we announced support?

Not sure if I am up to speed regarding this, or if I misunderstood the question.. We ship both client and server binaries and linux packages deb/rpms for AArch64, or `GOARCH=arm64` as it is also called. I guess that is official support, or did I miss something?

**Comment by odinuge** [comment]
Another viable solution is to fork runc and cherry-pick this commit https://github.com/opencontainers/runc/pull/2065/files#diff-d0f4afccf1d52087d9d4fd5ee1eccdf3R30, essentially just switching from `kB` to `KB`, and use that instead. If we think that "better" solution that we can live with. Then we don't have to bump all the dependencies, but still fix the issue.

As a side note, the same fix was indeed submitted into `runc` early 2015 https://github.com/opencontainers/runc/pull/149, but wasn't accepted. (but the "motivation" behind the fix was totally different, so I do kinda understand why it was closed.)

**Comment by odinuge** [comment]
Turns out `aws` have an offering of AArch64 nodes running ubuntu 19.04 and linux `5.0.0`, so I have done some testing.

## Running 1.15 deb from apt.kubernetes.io

Starting kubelet results in ths error:
```
Failed to start ContainerManager failed to initialize top level QOS containers: failed to update top level Burstable QOS cgroup : failed to set supported cgroup subsystems for cgroup [kubepods burstable]: Failed to set config for supported subsystems : failed to write 4611686018427387904 to hugetlb.64kB.limit_in_bytes: open /sys/fs/cgroup/hugetlb/kubepods/burstable/hugetlb.64kB.limit_in_bytes: permission denied
```
And kubelet exits with code 1

## Running Kubernetes v1.15.1-beta.0.21+26b85e575c97fc (this PR)

Kubelet starts normally, and reports the correct hugetlb capacity:

```
Capacity:
 cpu:                1
 ephemeral-storage:  7959716Ki
 hugepages-1Gi:      0
 hugepages-2Mi:      0
 hugepages-32Mi:     0
 hugepages-64Ki:     128Ki
 memory:             1889516Ki
 pods:               110
```
Reserved two pages using this command:

```
$ echo "2" > /sys/kernel/mm/hugepages/hugepages-64kB/nr_hugepages
```

Runnig a pod requesting `128Ki` with huge page memory with page size `64Ki` works as expected:

```
Containers:                                                                                                                                                                                   
  example:                                                                                                                                                                                    
    Container ID:  docker://043cf222b43e111de41760da7dc5e16458fb5b7da55150931359a33f1f89a0e1                                                                                                  
    Image:         alpine                                                                                                                                                                     
    Image ID:      docker-pullable://alpine@sha256:ca1c944a4f8486a153024d9965aafbe24f5723c1d5c02f4964c045a16d19dc54                                                                           
    Port:          <none>                                                                                                                                                                     
    Host Port:     <none>                                                                                                                                                                     
    Command:                                                                                                                                                                                  
      sleep
      12312321
    State:          Running
      Started:      Thu, 04 Jul 2019 22:24:01 +0000
    Ready:          True
    Restart Count:  0
    Limits:
      hugepages-64Ki:  128Ki
    Requests:
      hugepages-64Ki:  128Ki
      memory:          1Mi
    Environment:       <none>
    Mounts:
      /hugepages from hugepage (rw)
      /hugepages2 from hugepage2 (rw)
```

So this works as expected. Looks like we still have some discussion left, so ill just keep it on hold.

cc @mattjmcnaughton

**Comment by cblecker** [comment]
Okay, I reread things and I think I understand more clearly.. we ship stuff for AArch64, but it's the addition of linux 5.0+ that's causing this issue?

**Comment by mattjmcnaughton** [comment]
> Another viable solution is to fork runc and cherry-pick this commit https://github.com/opencontainers/runc/pull/2065/files#diff-d0f4afccf1d52087d9d4fd5ee1eccdf3R30, essentially just switching from `kB` to `KB`, and use that instead. If we think that "better" solution that we can live with. Then we don't have to bump all the dependencies, but still fix the issue. It will also lead to just a handful lines changed, making it way simpler to back port even further (if we decide to do so).
> 
> As a side note, the same fix was indeed submitted into `runc` early 2015 [opencontainers/runc#149](https://github.com/opencontainers/runc/pull/149), but wasn't accepted. (but the "motivation" behind the fix was totally different, so I do kinda understand why it was closed.)

Hmm, yeah I'm pretty hesitant to fork runc... seems like it could make things a lot more complex than we want them to be...

I'd be interested in the opinion of @dims, who has worked with dependency upgrades a lot in the past. And I'd also be interested in the opinion of the 1.15 branch manager, who I believe is @Bubblemelon :)

**Comment by odinuge** [comment]
> Hmm, yeah I'm pretty hesitant to fork runc... seems like it could make things a lot more complex than we want them to be...

Using a fork is pretty simple, but I do not know if that is what we want, but it is a real alternative to this huge dependency bump. Here is an example of how the "same back port" will look when using a fork: https://github.com/odinuge/kubernetes/pull/1 , with only +-7 lines.

> I'd be interested in the opinion of @_dims, who has worked with dependency upgrades a lot in the past. And I'd also be interested in the opinion of the 1.15 branch manager, who I believe is @_Bubblemelon :)

Smart!

**Comment by dims** [comment]
Please ... *NO* forks!! before we ask the release manager we should ask the sig node leads ( cc @derekwaynecarr @dchen1107 ) and also talk to @liggitt to see if this is appropriate.

my 2 cents, Document that folks wanting to use Aarch64 should apply https://github.com/opencontainers/runc/pull/2065/files#diff-d0f4afccf1d52087d9d4fd5ee1eccdf3R30 and rebuild their kubelet as a work around for now. Please note that unless we can set up CI(s) for this arch, we will never be able to ensure that things don't break for it. So i'd recommend figuring out how/where to do so as well if we really want to consider full-fledged support for this arch. At this point without CI, our commitment would be on a best-effort-basis. right?

**Comment by cblecker** [comment]
Yeah, my gut says that this is a "feature" as opposed to a "critical bug fix". If we haven't ever supported that combination of components before, then cherry picking back a fix isn't in like with our policy. Especially if we don't have CI coverage of this architecture.

**Comment by vielmetti** [comment]
@cblecker There is CI coverage at https://cncf.ci including arm64 systems.

**Comment by dims** [comment]
@vielmetti a few things need to line up before that's useful

- The problem being fixed by this cherry pick - was it seen in the CI? (FWIW, head is fixed and we are backporting here)
- Is the CI results being fed to test-grid?
- Will there be a CI test for 1.15 to help validate that the fix applied here?
- Is there a way to test this fix *before* we merge it?

So at this point, the CNCF CI is not a good indicator for signals for the release team IMHO.

thanks,
Dims

**Comment by cblecker** [comment]
Yeah, I'm not sure how the CNCF's CI system is set up.. it also appears that the ARM CI is failing (both at 1.13 stable and at "head"). From what I gather, this isn't strictly AArch64, but when a specific feature (HugePages) is enabled, which is the new default in Linux 5.0+.. this is why it wasn't previously seen.

Having reviewed the master PR, this cherry pick, and the discussions, what I see is this:
- A bug was introduced in 2017 when HugePages support was added in alpha (https://github.com/kubernetes/kubernetes/pull/50859)
- A similar bug was also in runc when it introduced hugetlb support in 2015 (https://github.com/odinuge/runc/commit/4002033269976cb3be4b5e8c370363105bede573)
- This runc bug was recently fixed (https://github.com/opencontainers/runc/pull/2065)..
- .. and then the Kubernetes dependency was bumped to pull in the updated runc, and Kubernetes code was updated to use the imported constant (https://github.com/kubernetes/kubernetes/pull/78495)
- Bumping runc also required bumping additional dependencies (github.com/checkpoint-restore/go-criu, github.com/sirupsen/logrus, and github.com/opencontainers/selinux which was bumped in a previous PR)
- These three dependency bumps would need to be picked back to 1.15 to bring in the fix (this PR), and may also require additional dependencies to be brought up to newer versions if it gets picked back further than 1.15.

Disturbing this much code for a const fix is unfortunate. I think it's the right fix in master, but I'm not convinced it's worth it to cherry pick back.

@odinuge Without this fix, are there other workarounds? Such as using a <5.0 Linux kernel? Or turning off HugePages? (not even sure the latter is an option)

**Comment by odinuge** [comment]
> Disturbing this much code for a const fix is unfortunate. I think it's the right fix in master, but I'm not convinced it's worth it to cherry pick back.

Totally agree! Just want some feedback from the sig-node guys about what they think. :smile: 

> Without this fix, are there other workarounds? Such as using a <5.0 Linux kernel? Or turning off HugePages? (not even sure the latter is an option)

Yes, there are. The user/distro distributor can compile their kernel without support for the hugetlb cgroup by setting `CGROUP_HUGETLB=n`. When we check if `hugetlb` is supported, we check for the hugetlb directory for the hugetlb cgroup.

The user can also temporary disable the cgroup (this applies to all kind of cgroups) by unmounting it. For most distros this can be done by running:

```bash
$ umount /sys/fs/cgroup/hugetlb
```

Both of these will however need some special actions from the users, so a third option is to disable the hugetlb support in kubelet if we detect a hugepage smaller than "1MB", and just log it to the user, instead of stopping kubelet.

https://github.com/kubernetes/kubernetes/blob/5b1e997fa393f2c71bdd937027360123efe4830e/pkg/kubelet/cm/cgroup_manager_linux.go#L328

Something like:
```go
hasUnsupportedPageSize := false
for _, pageSize := cgroupfs.HugePageSizes {
    if strings.HasSuffix(pageSize, "kB") {
        klog.Infof("Found unsupported hugetlb page size \"%s\", and therefore disables hugetlb support. Read more info on [GITHUB]\n", pageSize)
        // maybe link to issue on github with more info
        hasUnsupportedPageSize = true
    }
}
if !hasUnsupportedPageSize {
    supportedSubsystems[&cgroupfs.HugetlbGroup{}] = false
}
```

That way we don't require the user to do something special, but still allows kubelet to start.

**Comment by mattjmcnaughton** [comment]
> > Disturbing this much code for a const fix is unfortunate. I think it's the right fix in master, but I'm not convinced it's worth it to cherry pick back.
> 
> Totally agree! Just want some feedback from the sig-node guys about what they think. 
> 
> > Without this fix, are there other workarounds? Such as using a <5.0 Linux kernel? Or turning off HugePages? (not even sure the latter is an option)
> 
> Yes, there are. The user/distro distributor can compile their kernel without support for the hugetlb cgroup by setting `CGROUP_HUGETLB=n`. When we check if `hugetlb` is supported, we check for the hugetlb directory for the hugetlb cgroup.
> 
> The user can also temporary disable the cgroup (this applies to all kind of cgroups) by unmounting it. For most distros this can be done by running:
> 
> ```shell
> $ umount /sys/fs/cgroup/hugetlb
> ```
> 
> Both of these will however need some special actions from the users, so a third option is to disable the hugetlb support in kubelet if we detect a hugepage smaller than "1MB", and just log it to the user, instead of stopping kubelet.
> 
> https://github.com/kubernetes/kubernetes/blob/5b1e997fa393f2c71bdd937027360123efe4830e/pkg/kubelet/cm/cgroup_manager_linux.go#L328
> 
> Something like:
> 
> ```go
> hasUnsupportedPageSize := false
> for _, pageSize := cgroupfs.HugePageSizes {
>     if strings.HasSuffix(pageSize, "kB") {
>         klog.Infof("Found unsupported hugetlb page size \"%s\", and therefore disables hugetlb support. Read more info on [GITHUB]\n", pageSize)
>         // maybe link to issue on github with more info
>         hasUnsupportedPageSize = true
>     }
> }
> if !hasUnsupportedPageSize {
>     supportedSubsystems[&cgroupfs.HugetlbGroup{}] = false
> }
> ```
> 
> That way we don't require the user to do something special, but still allows kubelet to start.

Thanks for all the options! I'm intrigued by the last one... if I'm understanding everything correctly, the impacts of turning off `hugetlb` should be fairly small, and also only impact performance, correct? If yes, then I think one good option could be cherry-picking the fix to disable hugetlb, with the understanding that the niche of users who wanted to use Aarch64 and hugetlb would need to deploy a custom kubelet binary or wait for 1.16.

Thoughts?

**Comment by odinuge** [comment]
> if I'm understanding everything correctly, the impacts of turning off hugetlb should be fairly small, and also only impact performance, correct? If yes, then I think one good option could be cherry-picking the fix to disable hugetlb, with the understanding that the niche of users who wanted to use Aarch64 and hugetlb would need to deploy a custom kubelet binary or wait for 1.16.

The code snippet above will do the same as unmounting the hugetlb cgroup, but it would not "disable huge page support". Since the cgroup enforcement is separated from the capacity reporting in kubelet, the result will just be that the limits/requests will not be enforced, so all pods can use huge page memory until it is empty. This is implemented according to the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190129-hugepages.md#cgroup-enforcement), so if that is the right thing to do, is another discussion. A result of that, is that the a k8s user cannot know if the huge page requests/limits are enforced or not, unless they have access to the server. You can also risk that a pod is unable to use the amount it has requested.

If we are getting into the "monkeypatch" land, and if we are sure that runc will not be bumped in the future, we can just override the `HugePageSizes` variable inside `libcontainer` and then fix the actual problem instead. This would be quite easy to apply to other releases too. See here: https://github.com/kubernetes/kubernetes/compare/release-1.15...odinuge:hugetlb-fix-backport?expand=1

**Comment by kevin-zhaoshuai** [comment]
Kevin Zhao from Linaro. Thanks all for working on this. I've very interested about this cherry-pick and I can offer machines with Linux 5.0+ kernel for test. Will continue chasing on this

**Comment by dims** [comment]
@kevin-zhaoshuai as mentioned earlier (in https://github.com/kubernetes/kubernetes/pull/79671#issuecomment-508954097) we want a entry in our test-grid that the release team can use to signal if all the tests for a specific architecture works. We don't want an one off solution for one cherry pick. Can you please help with that?

**Comment by derekwaynecarr** [comment]
I think its fair to disable support for huge page sizes < 2MB in the 1.15 branch as that was the page size that was actually verified in e2e prior to the bug fix provided.  I apologize for the delay as I have been away from computer the past week.  If we can prepare a patch that ignores pages less than 2MB that seems a fine compromise and I am happy to approve that specific change.

**Review by derekwaynecarr — DISMISSED** [review]
Please prepare a smaller patch where the kubelet does not support page sizes less than 2Mi pending the fixes in 1.16.

**Comment by odinuge** [comment]
Hi @derekwaynecarr ! 😄 

> Please prepare a smaller patch where the kubelet does not support page sizes less than 2Mi pending the fixes in 1.16.

Hmm. That will require some "monkey-patching" inside both runc and the cadvisor-implementation.

Ref. my comment https://github.com/kubernetes/kubernetes/pull/79671#issuecomment-509025389, we can manually uppercase the value from runc, since we know that `HugePageSizes` in the current version is set only during startup, and if we update it before we use it, it should be ok. That will fix the problem, without disabling anything.

Here is the patch: https://github.com/kubernetes/kubernetes/compare/release-1.15...odinuge:hugetlb-fix-backport?expand=1

Any thoughts on that one @derekwaynecarr?

**Comment by kevin-zhaoshuai** [comment]
@dims, yes sure we can help for that and offer some resources for test. Now the problem is about how to setup the CI. Will ping you tomorrow on slack for more details.

**Comment by derekwaynecarr** [comment]
@odinuge 

I think for 1.15, we should just ignore page sizes < 2Mi, and I am thinking that should be as simple as editing the list of valid pageSizes in the `pageSizes` set here: 

see:
https://github.com/kubernetes/kubernetes/blob/release-1.15/pkg/kubelet/cm/cgroup_manager_linux.go#L400

As long as the page size is not in the HugeTlbLimit slice for values < 2Mi, I think there would be no issue on AArch64 (or any other location that had smaller huge pages).

The alternative change to edit the variable from runc would allow page sizes < 2Mi, but if its fine to say that one must wait to use page sizes < 2Mi for 1.16, that seems entirely reasonable to me and keeps the backport risk small.

To phrase another way, let's separate getting smaller page sizes from being able to start the kubelet on the desired architecture.

**Comment by odinuge** [comment]
Hmm. Okay. Guess that is reasonable. Still a bit concerned with the fact that removing support for sizes `< 1MiB` will end up with more mess than fixing it.

How do you want do that in practice? The internal code in `runc` that provide stats will still run for all the page sizes, and I guess it will crash with the same error as kubelet does now: https://github.com/kubernetes/kubernetes/blob/release-1.15/vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/hugetlb.go#L45

Then I guess we also should remove it from the reported capacity from cadvisor, here: https://github.com/kubernetes/kubernetes/blob/release-1.15/pkg/kubelet/cadvisor/util.go#L47

@derekwaynecarr

File: pkg/kubelet/cm/container_manager_linux.go
Comment (odinuge) [inline pkg/kubelet/cm/container_manager_linux.go]: I guess we can change this to `strings.ReplaceAll(pageSize, "kB", "KB")`to be even more explicit about what it does.

**Comment by odinuge** [comment]
Have you had a chance to look at my previous comment @derekwaynecarr? I have updated the patch with my suggestion (replacing kB with KB, so that all sizes are supported), and I have tested it, and it works well on both AArch64 and amd64.

---

Here is another implementation of disabling huge page sizes smaller than `1KiB`. Still think I am more fan of fixing it, as the current implementation does now, instead of this:

```diff
From fd36a49e208fb340b7b79211e2b9ca6d4651aea7 Mon Sep 17 00:00:00 2001
From: Odin Ugedal <odin@ugedal.com>
Date: Sun, 7 Jul 2019 21:37:54 +0200
Subject: [PATCH] Disable Huge pages smaller than 1MiB

---
 pkg/kubelet/cadvisor/util.go              |  3 +++
 pkg/kubelet/cm/cgroup_manager_linux.go    |  3 +++
 pkg/kubelet/cm/container_manager_linux.go | 14 ++++++++++++++
 3 files changed, 20 insertions(+)

diff --git a/pkg/kubelet/cadvisor/util.go b/pkg/kubelet/cadvisor/util.go
index 49e35fb5d0..43337fd1cd 100644
--- a/pkg/kubelet/cadvisor/util.go
+++ b/pkg/kubelet/cadvisor/util.go
@@ -45,6 +45,9 @@ func CapacityFromMachineInfo(info *cadvisorapi.MachineInfo) v1.ResourceList {
 
 	// if huge pages are enabled, we report them as a schedulable resource on the node
 	for _, hugepagesInfo := range info.HugePages {
+		if hugepagesInfo.PageSize < 1024 {
+			continue
+		}
 		pageSizeBytes := int64(hugepagesInfo.PageSize * 1024)
 		hugePagesBytes := pageSizeBytes * int64(hugepagesInfo.NumPages)
 		pageSizeQuantity := resource.NewQuantity(pageSizeBytes, resource.BinarySI)
diff --git a/pkg/kubelet/cm/cgroup_manager_linux.go b/pkg/kubelet/cm/cgroup_manager_linux.go
index 7797e9b0c2..1ee1c1f21b 100644
--- a/pkg/kubelet/cm/cgroup_manager_linux.go
+++ b/pkg/kubelet/cm/cgroup_manager_linux.go
@@ -389,6 +389,9 @@ func (m *cgroupManagerImpl) toResources(resourceConfig *ResourceConfig) *libcont
 	pageSizes := sets.NewString()
 	for pageSize, limit := range resourceConfig.HugePageLimit {
 		sizeString := units.CustomSize("%g%s", float64(pageSize), 1024.0, hugePageSizeList)
+		if strings.Contains(sizeString, "kB") {
+			continue
+		}
 		resources.HugetlbLimit = append(resources.HugetlbLimit, &libcontainerconfigs.HugepageLimit{
 			Pagesize: sizeString,
 			Limit:    uint64(limit),
diff --git a/pkg/kubelet/cm/container_manager_linux.go b/pkg/kubelet/cm/container_manager_linux.go
index a1f7ff1cc7..dd6fb794c7 100644
--- a/pkg/kubelet/cm/container_manager_linux.go
+++ b/pkg/kubelet/cm/container_manager_linux.go
@@ -200,6 +200,20 @@ func validateSystemRequirements(mountUtil mount.Interface) (features, error) {
 // Takes the absolute name of the specified containers.
 // Empty container name disables use of the specified container.
 func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.Interface, nodeConfig NodeConfig, failSwapOn bool, devicePluginEnabled bool, recorder record.EventRecorder) (ContainerManager, error) {
+	// Mitigation of the issue fixed in master where hugetlb prefix for page sizes with "KiB"
+	// is "kB" in runc, but the correct is "KB"
+	// See https://github.com/opencontainers/runc/pull/2065
+	// and https://github.com/kubernetes/kubernetes/pull/78495
+	// for more info.
+	var pageSizes []string
+	for _, pageSize := range fs.HugePageSizes {
+		if strings.Contains(pageSize, "kB") {
+			continue
+		}
+		pageSizes = append(pageSizes, pageSize)
+	}
+	fs.HugePageSizes = pageSizes
+
 	subsystems, err := GetCgroupSubsystems()
 	if err != nil {
 		return nil, fmt.Errorf("failed to get mounted cgroup subsystems: %v", err)
-- 
2.22.0
```

File: pkg/kubelet/cm/container_manager_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/cm/container_manager_linux.go]: I see how this is simpler than the alternative, and runc is making the variable global and writable for better or worse.  Since a proper fix was put in master, and there is no other known writer to that var I could find, this is probably reasonable and small.

**Comment by derekwaynecarr** [comment]
given all the options presented, i think this is probably the most reasonable path for a solution in the support stream with the smallest amount of code change.

thanks for your patience.

/lgtm
/approve

**Comment by odinuge** [comment]
This should be good to go now @kubernetes/patch-release-team! Looks like the @k8s-ci-robot is struggling a bit tho..

**Comment by odinuge** [comment]
This should be good to go now @kubernetes/patch-release-team! Looks like the @k8s-ci-robot is struggling a bit tho (probably because the github api is quite unstable atm.)..

**Comment by walduino** [comment]
Any chance this gets approved soon ? My Arm cluster is purely for educational purposes. I replaced one of the control plane nodes by an orange pi 3 sbc which runs linux 5.1.  I understand that there are only a few peeps waiting for it so there's little pressure. I CAN wait till september to start playing with k8s again, but would like not to HAVE TO. //cheerz

**Comment by Pennyzct** [comment]
Hi~ guys. I'm also working for ARM on Kata Containers, and need to build k8s cluster for it. The kernel version needs to be more than v5.0.x, so looks like v1.15.x is the best option. ;). really hope this PR get landed. ;)

**Comment by vielmetti** [comment]
Hoping that @deitch can take a peek at this and make a recommendation.

**Comment by deitch** [comment]
Trying to wrap my head around all of this. 

Basically the full fix is on `master`, targeted at 1.16. Because the full fix requires changing the dependencies we are very hesitant to backport it into 1.15 (I have been in the "dependency hell" of go modules too many times; I am disappointed not by go modules, but by how it tries to update and "magic things" a bit too much), so the fix that @derekwaynecarr said is "reasonable and small" is just to replace "kB" with "KB"? It won't properly fix the hugetlb issue as per #78495 does in `master`, but at least prevent it from crashing?

**Comment by tpepper** [comment]
Given the delays we had on the branch for the recent CVE's and that conversation on this cherry pick has continued a fair amount after it got lgtm/approve, can I get an affirmation from folks (eg: @derekwaynecarr @cblecker @dims @mattjmcnaughton) that ya'll do want this merged and things are complete.  We've begun merging content for 1.15.3 with a target of 2019-08-19.

**Review by cblecker — APPROVED** [review]
@tpepper Thanks for checking Tim. Appreciate your diligence, considering the background on this one.

Yes, I feel comfortable signing off on this one as the right approach for 1.15.

**Comment by dims** [comment]
+1 from me as well

**Pull request merged**
